Basic thread and domain structures for statmemprof

NickBarnes opened the pull request:
Statmemprof needs per-thread and per-domain data structures, and calls from the thread and domain modules to create and destroy them. This small PR adds skeleton structures and calls, without enabling statmemprof.

I am breaking the large statmemprof PR (#12379) into a number of small separate ones, some of which are wholly independent. This is one such.

NickBarnes committed:
Basic thread and domain interfaces to the memprof module, and data structures.

jhjourdan commented on runtime/domain.c:
```diff
@@ -1520,10 +1532,16 @@ void caml_interrupt_self(void) {
+  /* trigger at the greater of young_trigger and memprof_young_trigger */
+  value *trigger = dom_st->young_trigger;
+  if (dom_st->memprof_young_trigger &&
+      (dom_st->memprof_young_trigger > trigger)) {
```
If I remember well, in the previous implementation, the default value for `memprof_young_trigger` was `young_alloc_start` when memprof was not currently running.

It seems like you are in a somewhat redundant situation, where you both use `NULL` and `young_alloc_start`. Do I understand well? Any reason for this? I'm 100% fine with any situation (if I remember well, the choice of `young_alloc_start` was guided by the fact that the implementation was slightly simpler in this case), but it is weird to use both conventions.

jhjourdan replied:
(In any case, please document and add an assertion checking the possible values for `memprof_young_trigger`.)

NickBarnes replied:
Good catch. This was left-over from an earlier version of my multicore statmemprof code, in which memprof_young_trigger was only set if there was a memprof profile running. I'll simplify this away, and add an assert.

NickBarnes replied:
This has led to some rearrangement in domain initialization: the memprof domain state object now has to exist before the minor heap is created.

gadmm commented on otherlibs/systhreads/st_stubs.c:
```diff
@@ -360,7 +363,9 @@ static void caml_thread_remove_and_free(caml_thread_t th)
-/* TODO(engil): more work on the multicore fork machinery. */
+/* TODO(engil): more work on the multicore fork machinery.
+ * TODO(nick): in particular this only discards thread information for
+ * the current domain, not for all domains. */
```
This looks ok to me; this is in fact all domains as fork raises an exception if more than one domain is running.

jhjourdan commented on runtime/caml/domain_state.tbl:
```diff
@@ -15,9 +15,10 @@
-/* Minor heap limit. Typically young_limit == young_start, but this field is set
- * by other domains to signal this domain by causing a spurious allocation
- * failure. */
+
```
Spurious empty line.

jhjourdan commented on runtime/domain.c:
```diff
@@ -495,8 +501,10 @@ static int allocate_minor_heap(asize_t wsize) {
+  caml_memprof_renew_minor_sample(domain_state);
   caml_reset_young_limit(domain_state);
 
+
```
Spurious empty line.

jhjourdan commented on runtime/memprof.c:
```diff
@@ -15,6 +15,276 @@
+  memprof_thread_t thread = domain->current;
+
+  return (thread
+          && !atomic_load(&domain->config.stopped)
+          && domain->config.lambda > 0
```
What's the semantics of `lambda == 0` if Memprof is stopped in the current domain?

NickBarnes replied:
This is an excellent question. This is descended from a similar test in the old statmemprof. I think probably we should raise an exception in `start` if lambda == 0 (as we do for lambda < 0 or lambda > 1). In the old statmemprof, starting with zero lambda does not raise an exception but (a) all sampling is short-circuited off (rather than simply never occurring), and (b) stopping is done, in part, by setting lambda to zero. New statmemprof has more explicit stopping mechanisms (and more complex, due to multicore), so using lambda for this could go away.

jhjourdan replied:
Ok, let's do that, then.

jhjourdan commented on runtime/memprof.c:
```diff
@@ -15,6 +15,276 @@
+
+CAMLexport memprof_thread_t caml_memprof_new_thread(caml_domain_state *domain)
+{
+  CAMLassert(domain->memprof);
```
If I understand well, that assertion will fail if the creation of Memprof's domain structure failed.

Frankly, we should rather raise OOM if allocation of such structures fails.

NickBarnes replied:
I agree (although I'm not sure what happens when a domain-startup allocation fails; it's not clear from a brief check of domain.c).

NickBarnes replied:
I've looked into this a bit more deeply and rearranged some of the code in `domain_create` so that we don't set `domain_self` if either this or the `caml_domain_state` allocation fails, which both callers of that function now turn into a `caml_fatal_error`. That is at least consistent; possibly `Domain.spawn` should raise an exception rather than a fatal error, but there are other ways for that to get `caml_fatal_error` (e.g. "Failed to create domain: signal stack") so that is also more consistent than previously.

jhjourdan replied:
Good!

jhjourdan commented on runtime/memprof.c:
```diff
@@ -15,6 +15,276 @@
+
+struct memprof_config_s {
+  /* Whether memprof has been stopped. */
+  bool _Atomic stopped;
```
I guess this will become clear in a later PR, but why is this atomic? Is this because of the profile-sharing system, where a domain can be stopped by another one?

NickBarnes replied:
Yes, exactly.

jhjourdan reviewed (approved):
I have a few minor comments, but all in all this looks good so far.

gadmm commented on runtime/domain.c:
```diff
@@ -297,6 +298,11 @@ Caml_inline void interrupt_domain(struct interruptor* s)
 
+Caml_inline dom_internal *internal_of_interruptor(struct interruptor *s)
+{
+  return (dom_internal *)((char *)s - offsetof(dom_internal, interruptor));
```
I think this calls for a bigger reorganization to avoid such tricks. Have you considered other ways to make the parent domain accessible at domain creation?

NickBarnes replied:
I am used to this `offsetof` technique from other work but I'm happy to do it another way. The easy alternative is to put the `caml_domain_state *` into the `domain_startup_params`.

NickBarnes replied:
Or just the whole `dom_internal *` as `parent`, rather than two of its fields. Simplifies some code.

gadmm replied:
I dug deeper as I was a bit puzzled as to why one maintains a pointer to another domain's state when access to this domain state is usually local-only for thread-safety reasons. This is where the `_Atomic stopped` in `memprof_config_s` comes into play as it is meant to be accessed in parallel.

My first worry is the following: while the domain state is allocated once and persists through domain reallocation (this is how `young_limit` can be changed from another domain without worries about memory safety), here the `memprof_config_s` (via `memprof_domain_s`) is allocated dynamically and deallocated upon domain shutdown. This makes me nervous about the thread-safety of accessing `stopped` from another domain. (To make things simpler, it seems that one could allocate the domain-local memprof data once and for all and recycle the allocated structure, like one does for the domain state.)

My second observation is that the memprof runtime has a concept of a parent-child relationship between domains whereas the domain runtime itself does not maintain such a relationship. This might explain why the code is a bit convoluted. I wonder how this parent-child relationship is important API-wise; it seems that only the relationship of having the same memprof "profile" is important. (Maybe I am reading too much into the variable names and this idea of parent-child does not affect the rest of the implementation.)

My last worry is that you use this parent domain state only (IIUC) at domain creation in order to duplicate the `stopped` mutable status and the config parameters. I am wondering how much things could be simpler if this atomic `stopped` switch was in fact on the OCaml heap and shared between domains (avoiding state replication). Pushing this idea further, could the `profile` consist of a heap-allocated OCaml value (comprising the `stopped` boolean and the rest of the data)? It looks to me as if this organisation could simplify things (and it will make sense to pass this profile value among the startup params). Do you have an insight to share that would clarify why things are done as they are now (notably regarding state replication of the `stopped` flag)?

NickBarnes replied:
The parent/child relationship does not persist in the memprof data structures; it is only used to initialize memprof status for the new domain (by copying that from the parent domain).

NickBarnes replied:
More detail: the memprof data structure for a child domain does not keep any reference to the parent domain's data structures. The configuration is copied, not pointed to. Memprof *will* keep track of the overall set of per-domain memprof structures, and will sometimes have to traverse this overall structure (for example, in order to set the per-domain `stopped` flag for one or more other domains). Such traversal and setting will be guarded by appropriate locks etc, for thread safety.

The reason for using a per-domain `stopped` flag, rather than a per-profile flag, is that it is checked frequently by its owning domain, only set at all when the profile is stopped, and only set by another domain when a profile is stopped which is shared between several domains. So in (what I think of as) the common case - in which domains do not share profiles - there should be very low cost to the atomicity (a domain's config can live in that domain's core's cache with no cache consistency traffic).

The idea of having a per-profile object on the Caml heap did occur to me. I didn't see a need for it, but we could migrate to such a system in the future, with only very localised code changes.

gadmm commented on runtime/printexc.c:
```diff
@@ -143,6 +143,12 @@ void caml_fatal_uncaught_exception(value exn)
+     a memprof callback could raise an exception while
+     [handle_uncaught_exception] is running, and the printing of
+     the exception could fail. */
+  caml_memprof_update_suspended((_Bool)1);
```
Let's include `stdbool.h` and use `true`, there is no good point in using `_Bool` otherwise.

gadmm reviewed (commented):
See below for comments.

Regarding error handling in domain creation, the change looks good to me as it currently stands.

gadmm commented on runtime/domain.c:
```diff
@@ -1119,32 +1134,45 @@ static void* domain_thread_func(void* v)
+  /* Handshake completed: now we can fail! */
+  if (failure) {
+    caml_fatal_error("%s", failure);
+  }
```
As `caml_fatal_error` will stop the program entirely, there is no point in awaiting the completion of the handshake.

NickBarnes replied:
There has been discussion of surviving failures in `Domain.spawn`, with an exception. If we ever do that, this can become an exception-raising path.

gadmm replied:
You might be referring to #12410. Then do not worry and let me handle it :) For now the smallest rebase conflict the simpler it is for me.

NickBarnes replied:
OK, I've backed out that control-flow change.

NickBarnes commented:
Rebased to trunk. I think I've resolved everything from @gadmm and @jhjourdan

gadmm commented:
I am still wondering about the first and third points of https://github.com/ocaml/ocaml/pull/12382#discussion_r1282305491. I'll have more time to look at the code later this month, but can you tell me how you addressed the first one?

As for the third one, note that a shared atomic flag is not more expensive than a local flag when the flag is updated rarely (it stays in cache just like a local flag). So I am not convinced by the stopped flag being checked frequently as a good reason for using a per-domain flag rather than per-profile.

The "push" architecture with duplication of state makes things quite complicated, for instance you have a complicated cross-domain synchronization inside `caml_memprof_discard`. IIUC, with a "pull" architecture with shared data, the discard status could simply be an additional shared flag which one checks before running the callbacks (this would also remove the corner-case where you raise an exception if the profile has not been stopped). Importantly, having a GC is handy to trivialize the collection of this shared state.

NickBarnes commented:
I've rebased and reworked this, based on @gadmm's suggestions, working towards a new complete statmemprof implementation in which the configuration is kept in a Caml heap object. This does introduce an indirection, but until we have evidence that this has a measurable performance impact I concur with @gadmm that this is cleaner. In the complete implementation, the config heap object can also stand as the `Gc.Memprof.t` value.

NickBarnes commented:
rebased.

NickBarnes commented:
Rebased again.

sadiqj reviewed (approved):
Took some time bouncing between `domain_terminate` and `domain_create` to convince myself that the logic was right but I think I'm happy.

sadiqj commented:
Thanks @NickBarnes !
